### PR TITLE
Support  service discovery with SRV record

### PR DESCRIPTION
**Pull request by ganmacs**
**Which issue(s) this PR fixes**: 
follow up https://github.com/fluent/fluentd/pull/2496

**What this PR does / why we need it**: 

**Docs Changes**:

needed

**Release Note**: 

Same as the title

### CODE REVIEW

File: lib/fluent/plugin/sd_srv.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+          queue.push(a)
+        end
+      end
+
+      def fetch_srv_record
+        adders =
+          begin
+            @dns_resolve.getresources(@target, Resolv::DNS::Resource::IN::SRV)
+          rescue => e
+            raise e               # TODO
```
Comment (repeatedly) [inline lib/fluent/plugin/sd_srv.rb]: What is this TODO?

File: lib/fluent/plugin/sd_srv.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        end
+
+        services.sort_by(&:first).flat_map { |s| s[1] }
+      end
+
+      def dns_lookup!(host)
+        # may need to cache the result
+        @dns_resolve.getaddress(host) # get first result for now
+      rescue => e
+        raise e             # TODO
```
Comment (repeatedly) [inline lib/fluent/plugin/sd_srv.rb]: ditto

File: lib/fluent/plugin/sd_srv.rb (hunk 1 above)
Comment (ganmacs) [reply lib/fluent/plugin/sd_srv.rb]: Never mind.  I thought it needed to rescue and output something log. but it's done in refresh_srv_records.

https://github.com/fluent/fluentd/pull/2876/commits/e7d82e05c311e381709100e4a6c88d080f96d675

**Comment by ganmacs** [comment]
@repeatedly @cosmo0920 

I added some config to support password, username, shared_key. Could you take a look at it? 
https://github.com/fluent/fluentd/pull/2876/commits/93ca09e573e11bfcfd78009e5247e10a2a8ae7e8

File: test/plugin/test_sd_srv.rb [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+      mock = flexmock('dns_resolver', getaddress: '127.0.0.1')
+               .should_receive(:getresources).with("_service1._udp.example.com", Resolv::DNS::Resource::IN::SRV)
+               .and_return([SRV_RECORD1, SRV_RECORD2])
+               .mock
+
+      mock(Resolv::DNS).new { mock }
+      sdf.configure(config_element('service_discovery', '', { 'service' => 'service1', 'hostname' => 'example.com', 'proto' => 'udp' }))
+      assert_equal Fluent::Plugin::ServiceDiscovery::Service.new(:srv, '127.0.0.1', 8081, 'service1.example.com', 10, false, '', '', nil), sdf.services[0]
+      assert_equal Fluent::Plugin::ServiceDiscovery::Service.new(:srv, '127.0.0.1', 8082, 'service2.example.com', 20, false, '', '', nil), sdf.services[1]
+    end
```
Comment (cosmo0920) [inline test/plugin/test_sd_srv.rb]: `shared_key`, `username`, and `password` testcases are missing.
Should we add testcases for them?

Comment (ganmacs) [reply test/plugin/test_sd_srv.rb]: added https://github.com/fluent/fluentd/pull/2876/commits/42caa1847a6e528c3efcae381c573d1abfae4607

**Pull request merged**